### ISSUE TITLE
Tpetra:  Increased memory usage with latest Trilinos

### ISSUE THREAD

**Issue by kddevin** [trilinos/Trilinos#6648]
## Bug Report
@trilinos/tpetra 

### Description
From @crdohrm 
We've noticed a large increase in the memory high water mark for some of our tests. This morning we isolated it to some doImport operations on Tpetra CrsMatrices. For the case below, we have a fillCompleted CrsMatrix inputMatrix and would like to extract a subset of its rows and columns into outputMatrix (see function below).

 ```
getTpetraMatrix(RCP<const CrsMatrix> inputMatrix,
            RCP<const Map> RowMap,
            RCP<CrsMatrix> & outputMatrix) const
{
  RCP<const Map> RowMap1to1 = inputMatrix->getRowMap();
  Import Importer(RowMap1to1, RowMap);
  outputMatrix = rcp( new CrsMatrix(RowMap, RowMap, 0) );
  outputMatrix->doImport(*inputMatrix, Importer, Tpetra::INSERT);
  outputMatrix->fillComplete(RowMap1to1, RowMap1to1);
}
```

Some data for the existing version of Trilinos in Sierra follows.

Memory high water mark in MB (max over processes) for existing Trilinos in Sierra:
```
just before doImport: 688
just before fillComplete: 720
just after fillComplete: 822
```

The corresponding numbers using the new version of Trilinos:
```
just before doImport: 712
just before fillComplete: 1424
just after fillComplete: 1622
```
 
Anyway, we just wanted to know if you see anything obvious in the function above that could be done to avoid the large increase in memory. We tried constructing a CrsGraph first and then use that in the constructor for outputMatrix, but saw some errors during fillComplete of the CrsGraph (we think these errors may be related to using the same row and column map in the constructor).

Here are some more old/new Trilinos results for some other tests (notice Max memory high water):

old Trilinos:
```
PERFORMANCE TEST 1
Parallel Memory high water (MB):  Min: 180.3  Max: 360.1  Avg: 296.4  Sum: 2371.2

PERFORMANCE TEST 2
Parallel Memory high water (MB):  Min: 146.8  Max: 217.7  Avg: 183.2  Sum: 1465.7
 ```

new Trilinos:
```
PERFORMANCE TEST 1
Parallel Memory high water (MB):  Min: 241.4  Max: 847.1  Avg: 672.6  Sum: 5380.7

PERFORMANCE TEST 2
Parallel Memory high water (MB):  Min: 208.7  Max: 368.7  Avg: 289.7  Sum: 2317.9
```
### Steps to Reproduce
1. SHA1:  [insert here]
1. Configure script:  [attach here]
1. Configure log:  [attach here]
1. Build log:  [attach here]
1. Input deck:  [attach here]
1. Do this.
1. Do that.
1. Shake fist angrily at computer.
1. Run log:  [attach here]

**Comment by shardest** [comment]
Quoting a previous e-mail:

We stepped through the new Trilinos doImport with the debugger to see where the large memory allocation happened:
 
Tpetra_DistObject_def.hpp: 911 (packAndPrepare - noticeable increase)
Tpetra_DistObject_def.hpp: 1237 (unpackAndCombine - huge increase)
 
The large increased allocation we see is primarily occurring inside unpackAndCombine

**Comment by mhoemmen** [comment]
@crdohrm wrote:
> (we think these errors may be related to using the same row and column map in the constructor).

It looks like you intend `outputMatrix` to be a possibly overlapping subdomain matrix.  Ifpack2 handles those in a different, weird, but somehow less troublesome way.  @csiefer2 knows a bit about that.  In any case, there's nothing wrong with using the same row and column Map in the constructor -- in fact, that should use less memory overall, since you're always storing local indices by default (unless you do something weird like fill locally with global indices).

**Comment by mhoemmen** [comment]
@shardest Are you possibly doing nonlocal inserts?  That is, do you ever call `insert{Global,Local}Indices` on a `Tpetra::CrsGraph` with a row index that is not in the graph's row Map on the calling process?

**Comment by crdohrm** [comment]
I don't recall doing any nonlocal inserts. Pretty much what we're doing is summarized by the function shown at the top of Description.

**Comment by shardest** [comment]
Recalling the max memory usage numbers from above - the first performance test high water mark went from 360 MB to 847 MB.

I'm looking at the implications of the double copying described in #6663 : if we assume that both tests have a baseline memory usage of b, and a memory usage m for the matrix, then we can estimate the effect of reducing three copies to two:

[360; 847] = [1, 1; 1, 3]*[b; m]
[b; m] = [116.5; 243.5]

b + 2*m = 603.5, still 67% more than the older version of Trilinos. That would be an improvement, but would still cause severe usability problems. Are these assumptions reasonable?

Just wanted to throw this out there in case we also need to consider a different approach to handling this problem in SD.

**Comment by mhoemmen** [comment]
@shardest Thanks for doing the estimation!  The main issue is whether the extra 67% would block Trilinos integration.  Would this cause the performance tests to fail?

I'm pretty sure Tpetra could further reduce peak memory use for this use case ({Ex,Im}port into an empty CrsMatrix), but it would take quite a bit longer than just fixing the 3x duplication.  Trilinos integration is becoming an urgent issue, so I'd like to know if this is absolutely a blocker for SD.

**Comment by mhoemmen** [comment]
@shardest @crdohrm Another thing you could try as a work-around, is to call [importAndFillCompleteCrsMatrix](
https://github.com/trilinos/Trilinos/blob/c962db7665df7f841f62de8f3f7058e0ba6c3449/packages/tpetra/core/src/Tpetra_CrsMatrix_decl.hpp#L4568Z).  It's supposed to be faster anyway.
```cpp
RCP<CrsMatrix> outputMatrix = 
  Tpetra::importAndFillComplete(inputMatrix, importer, domainMap, rangeMap);
```
The result is fill complete.

**Comment by mhoemmen** [comment]
Update:
1. `importAndFillCompleteCrsMatrix` won't work here, because SD specifies the target's column Map explicitly.  (It happens to equal the target's row Map.)  The "domain Import" argument is an Import from the source domain Map to the target domain Map; that's not what we want here.
2. I verified that `pad_crs_arrays` was getting called.  However, I haven't yet verified that this is the source of the large memory allocations.  I will do that next.

We had a lot of trouble with debuggers today, so my plan is to verify allocations manually using print statements.

**Comment by khpierson** [comment]
From my perspective, this issue is a blocker for a Trilinos integration. The potential impact on existing SD production analysis is too risky. -Kendall

**Comment by mhoemmen** [comment]
@crdohrm wrote:
> I don't recall doing any nonlocal inserts.

@shardest and I verified today that SD wasn't doing any nonlocal inserts, at least not in the function call that caused the test to run out of memory.

**Comment by nate945** [comment]
Commenting so that I get emails related to the ticket.  Thanks!

**Comment by mhoemmen** [comment]
@nate945 I just sent you, @crdohrm, and @shardest an e-mail with the log file and instructions for reproducing the experiment.

**Comment by srajama1** [comment]
@srajama1 : Adding myself to get updates

**Comment by nate945** [comment]
Thanks for the example. I did some data gathering on that case Mark sent using the currently installed Sierra version of Trilinos (Trilinos from ~June 2019).  The static profile seems to have the same high memory behavior in our currently installed Trilinos version.   Thus, with the installed version can evaluate the effect of the profile options in that one routine in isolation.
Gathered Data (Structured Hex20 mesh, 8 procs, ~20,000 elements)


Tpetra Matrix Constructor Option for overlap preconditioner matrix | Post - doImport: Increase in memory in use | Post - doImport: Increase in memory high water mark | Total Sierra/SD Memory use (sum all ranks)
-- | -- | -- | --
CrsMatrix(RowMap, RowMap, 0, DynamicProfile) | 357 Mb | 357 Mb | 6631 Mb
CrsMatrix(RowMap, RowMap, 0, StaticProfile) | 210 Mb | 1731 Mb | 13950 Mb
CrsMatrix(RowMap, RowMap, 243, StaticProfile) | 212 Mb | 1704 Mb | 13840 Mb

The first test case is the dynamic profile constructor, the is the default matrix construction option in our installed (June 2019) Trilinos.  Next, I tired setting a static profile with row size of zero (which doImport seems to handle just fine).  Also tried setting a conservative upper bound on row size in the constructor (this is a structured hex mesh with 81 nodes in the stencil, 3 dofs per node, and thus a max of 243 entries per row).    
The import to the static matrix seems to incur some sort of huge temporary scratch space overhead.  I’ve checked this effect on a variety of different problem sizes, element types, processor counts, etc. and this big memory hit seems to be present across the board.  In general, Sierra/SD is very likely to exceed total memory available and crash during this import step.  Sierra/SD is often run very near the processor memory limits in normal production use.
On the plus side the static profile uses less permanent memory.  Presumably the dynamic profile is adding some sort of growth padding.  To answer Mark and Sean’s previous question, if the import scratch space overhead was limited to 2X of the matrix size that may well be good enough to resolve this.  Dynamic profile was already burning some memory anyway.  However, current behavior seems to more like an 8X+ scratch memory overhead during import, which is often acutely fatal, and algorithmically seems very surprising.
As an aside it was very baffling to me how SD was even using dynamic profile in the first place.  The dynamic profile option is never requested in any Sierra side code, has been deprecated for over a year, and I don’t think any of the SD algorithms should require it.  It looks like in our current (June 2019) Trilinos version there is a line in Tpetra_ConfigDefs.hpp that sets the default profile type to “StaticProfile+1” which apparently means dynamic profile.  SD does indeed use the default profile matrix constructor in some places.  Looks to me that in the Sierra build system may silently enable deprecated features right up to the point they are deleted, and then of course everything breaks, seemingly out of the blue.  Maybe this is old news to others, but I found it very enlightening.  Might be something to follow up on here.

**Comment by mhoemmen** [comment]
@nate945 wrote:
> However, current behavior seems to more like an 8X+ scratch memory overhead during import, which is often acutely fatal, and algorithmically seems very surprising.

FYI here is a likely explanation for the memory increase: https://github.com/trilinos/Trilinos/issues/6663#issuecomment-580864613

**Comment by mhoemmen** [comment]
@nate945 wrote:
> As an aside it was very baffling to me how SD was even using dynamic profile in the first place.

That was Tpetra's default behavior before, and had always been.  That was also Epetra's default behavior (Epetra also has this DynamicProfile / StaticProfile option).

**Comment by mhoemmen** [comment]
I'm calling this closed, since we fixed #6663.  Please feel free to reopen if this is still an issue.  Thanks all!!! :-D
